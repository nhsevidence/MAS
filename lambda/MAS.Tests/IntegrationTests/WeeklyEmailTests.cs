using MailChimp.Net.Core;
using MAS.Configuration;
using MAS.Controllers;
using MAS.Services;
using MAS.Tests.Fakes;
using MAS.Tests.Infrastructure;
using Microsoft.Extensions.Logging;
using Moq;
using Shouldly;
using System;
using System.Threading.Tasks;
using Xunit;
using Xunit.Abstractions;

namespace MAS.Tests.IntegrationTests
{
    public class WeeklyEmailTests : TestBase
    {
        public WeeklyEmailTests(ITestOutputHelper output) : base(output)
        { }

        [Fact]
        public async void EmailBodyHtmlMatchesApprovedWithMEC()
        {
            // Arrange
            ContentRequest contentRequest = null;
            var fakeMailChimpManager = new FakeMailChimpManager();
            fakeMailChimpManager
                .Setup(x => x.Content.AddOrUpdateAsync(It.IsAny<string>(), It.IsAny<ContentRequest>()))
                .Callback<string, ContentRequest>((cId, conReq) => contentRequest = conReq);

            var client = _factory
                .WithImplementation(fakeMailChimpManager.Object)
                .WithCMSConfig(cmsConfig => cmsConfig.WeekliesBySendDatePath = "/weekly.json")
                .CreateClient();

            // Act
            await client.PutAsync("/api/mail/weekly/2020-01-13", null);

            // Assert
            ((string)contentRequest.Template.Sections["body"]).ShouldMatchApproved();
        }

        [Fact]
        public async void EmailBodyHtmlMatchesApprovedWithoutMEC()
        {
            // Arrange
            ContentRequest contentRequest = null;
            var fakeMailChimpManager = new FakeMailChimpManager();
            fakeMailChimpManager
                .Setup(x => x.Content.AddOrUpdateAsync(It.IsAny<string>(), It.IsAny<ContentRequest>()))
                .Callback<string, ContentRequest>((cId, conReq) => contentRequest = conReq);

            var client = _factory
                .WithImplementation(fakeMailChimpManager.Object)
                .WithCMSConfig(cmsConfig => cmsConfig.WeekliesBySendDatePath = "/weekly-without-mec.json")
                .CreateClient();

            // Act
            await client.PutAsync("/api/mail/weekly/2020-01-13", null);

            // Assert
            ((string)contentRequest.Template.Sections["body"]).ShouldMatchApproved();
        }

        [Fact]
        public async Task GivenNormalMonday_SendEmail()
        {
            //Arrange
            var mockContentService = new Mock<IContentService>();
            var mailController = new MailController(new FakeMailChimpManager().Object, Mock.Of<IMailService>(), mockContentService.Object, Mock.Of<IViewRenderer>(), new FakeBankHolidayService(), Mock.Of<ILogger<MailController>>(), Mock.Of<MailConfig>(), TestAppSettings.MailChimp.Default);

            var date = new DateTime(2020, 1, 13);

            //Act
            await mailController.PutWeeklyMailAsync(date);

            // Assert
            mockContentService.Verify(mock => mock.GetWeeklyAsync(date), Times.Once());
        }

        [Fact]
        public async Task GivenTuesdayAndPreviousMondayWasNotBankHoliday_EmailWasAlreadySent()
        {
            //Arrange
            var client = _factory
              .CreateClient();

            //Act
            var response = await client.PutAsync("/api/mail/weekly/2020-01-14", null);
            var responseContent = await response.Content.ReadAsStringAsync();

            // Assert
            responseContent.ShouldBe("An email was sent on 13/01/2020");
        }

        [Fact]
        public async Task GivenMondayIsBankHoliday_DontSendEmail()
        {
            //Arrange
            var client = _factory
              .CreateClient();

            //Act
            var response = await client.PutAsync("/api/mail/weekly/2019-08-26", null);
            var responseContent = await response.Content.ReadAsStringAsync();

            // Assert
            responseContent.ShouldBe("26/08/2019 is a bank holiday therefore an email isnt sent");
        }

        [Fact]
        public async Task GivenTuesdayAndPreviousMondayWasBankHoliday_SendEmail()
        {
            //Arrange
            var mockContentService = new Mock<IContentService>();
            var mailController = new MailController(new FakeMailChimpManager().Object, Mock.Of<IMailService>(), mockContentService.Object, Mock.Of<IViewRenderer>(), new FakeBankHolidayService(), Mock.Of<ILogger<MailController>>(), Mock.Of<MailConfig>(), TestAppSettings.MailChimp.Default);

            var date = new DateTime(2019, 8, 27);
            var previousMonday = new DateTime(2019, 8, 26, 00, 00, 00);

            //Act
            await mailController.PutWeeklyMailAsync(date);

            // Assert
            mockContentService.Verify(mock => mock.GetWeeklyAsync(previousMonday), Times.Once());
        }
    }
}
