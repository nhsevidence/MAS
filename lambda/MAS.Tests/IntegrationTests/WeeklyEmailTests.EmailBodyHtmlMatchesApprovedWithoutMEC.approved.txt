
<div>


<table class="spacer-heading">
    <tbody>
        <tr>
            <td height="16px" style="font-size:16px; line-height:16px;">
                <h3 class="article-heading">
                    Guidance and advice
                </h3>
            </td>
        </tr>
    </tbody>
</table>
<table class="row article-border">
    <tbody>
        <tr>
            <th class="small-12 large-12 columns first last">
                <table>
                    <tbody>
                        <tr>
                            <th>
                                <h4 class="h4 article-title">
                                        <a class="article-link" href="https://www.medicines.org.uk/emc/product/10478/smpc">
                                            Emgality (galcanezumab) injection in pre-filled pen 
                                        </a>
                                </h4>
                                <p class="article-description">
                                    SPC now advises that serious hypersensitivity reactions including cases of anaphylaxis, angioedema and urticaria have been reported. If a serious hypersensitivity reaction occurs, administration of galcanezumab should be discontinued immediately and appropriate therapy initiated
                                </p>
                                <p class="article-tags">
                                    <span class="source-tag">
                                        electronic Medicines compendium
                                    </span>
                                        <span class="specialism-tag">
                                            Allergy and immunology
                                        </span>
                                        <span class="specialism-tag">
                                            Emergency medicine and urgent care
                                        </span>
                                        <span class="specialism-tag">
                                            Neurological disorders
                                        </span>
                                </p>
                            </th>
                            <th class="expander"></th>
                        </tr>
                    </tbody>
                </table>
            </th>
        </tr>
    </tbody>
</table>
<table class="row article-border">
    <tbody>
        <tr>
            <th class="small-12 large-12 columns first last">
                <table>
                    <tbody>
                        <tr>
                            <th>
                                <h4 class="h4 article-title">
                                        <a class="article-link" href="https://www.thelancet.com/journals/lanpsy/article/PIIS2215-0366(18)30471-1/fulltext">
                                            Opioid prescribing trends and geographical variation in England
                                        </a>
                                </h4>
                                <p class="article-description">
                                    There were substantial increase in opioid (OP) prescribing in NHS between 1998 and 2016; OP prescriptions increased by 34% (from 568/1000 patients to 761/1000). After correcting for total oral morphine equivalency, increase was 127% (from 190,000 to 431,000mg/1000 population).
                                </p>
                                    <p class="article-description">
                                        <a class="article-link-subtle standard-link" href="http://mas-integration-tests/opioid-prescribing-trends-and-geographical-variation-in-england.html" aria-label="Read SPS commentary for Opioid prescribing trends and geographical variation in England">
                                            Read SPS commentary
                                        </a>
                                    </p>
                                <p class="article-tags">
                                    <span class="source-tag">
                                        The Lancet Psychiatry
                                    </span>
                                        <span class="specialism-tag">
                                            Anaesthesia and pain
                                        </span>
                                        <span class="specialism-tag">
                                            Mental health and illness
                                        </span>
                                </p>
                            </th>
                            <th class="expander"></th>
                        </tr>
                    </tbody>
                </table>
            </th>
        </tr>
    </tbody>
</table>
<table class="row article-border">
    <tbody>
        <tr>
            <th class="small-12 large-12 columns first last">
                <table>
                    <tbody>
                        <tr>
                            <th>
                                <h4 class="h4 article-title">
                                        <a class="article-link" href="https://jamanetwork.com/journals/jama/fullarticle/2759414">
                                            Effect of Vitamin C, Hydrocortisone, and Thiamine vs Hydrocortisone Alone on Time Alive and Free of Vasopressor Support Among Patients With Septic Shock
                                        </a>
                                </h4>
                                <p class="article-description">
                                    Exploratory RCT (n=216) found IV vitamin C, hydrocortisone &amp; thiamine vs. IV hydrocortisone alone, does not significantly improve duration of time alive &amp; free of vasopressor administration over 7 days (122.1 hrs vs. 124.6 hrs, respectively).
                                </p>
                                    <p class="article-description">
                                        <a class="article-link-subtle standard-link" href="http://mas-integration-tests/effect-of-vitamin-c-hydrocortisone-and-thiamine-vs-hydrocortisone-alone-on-time-alive-and-free-of-vasopressor-support-among-patients-with-septic-shock.html" aria-label="Read SPS commentary for Effect of Vitamin C, Hydrocortisone, and Thiamine vs Hydrocortisone Alone on Time Alive and Free of Vasopressor Support Among Patients With Septic Shock">
                                            Read SPS commentary
                                        </a>
                                    </p>
                                <p class="article-tags">
                                    <span class="source-tag">
                                        Journal of the American Medical Association
                                    </span>
                                </p>
                            </th>
                            <th class="expander"></th>
                        </tr>
                    </tbody>
                </table>
            </th>
        </tr>
    </tbody>
</table>
</div>
